waku/v2/waku_archive/driver/postgres_driver/postgres_driver.nim

```diff
@@ -0,0 +1,288 @@
+                                   toHex(message.payload),
+                                   pubsubTopic,
+                                   int64(message.version),
+                                   message.timestamp)
+    if not res:
+      return err("failed to insert into database")
+  except DbError:
+    return err("failed to insert into database")
+
+  return ok()
+
+proc extractRow(r: Row): ArchiveDriverResult[ArchiveRow] =
+    var wakuMessage: WakuMessage
+    var timestamp: Timestamp
+    var version: uint
+    var pubSubTopic: string
+    var contentTopic: string
+    var storedAt: int64
+    var digest: string
+    var payload: string
+
+    try:
+      storedAt = parseInt(r[0])
+      contentTopic = r[1]
+      payload = parseHexStr(r[2])
+      pubSubTopic = r[3]
+      version = parseUInt(r[4])
+      timestamp = parseInt(r[5])
+      digest = parseHexStr(r[6])
+    except ValueError:
+      return err("could not parse timestamp")
+
+    wakuMessage.timestamp = timestamp
+    wakuMessage.version = uint32(version)
+    wakuMessage.contentTopic = contentTopic
+    wakuMessage.payload = @(payload.toOpenArrayByte(0, payload.high))
+
+    ok((pubSubTopic,
+        wakuMessage,
+        @(digest.toOpenArrayByte(0, digest.high)),
+        storedAt))
+
+method getAllMessages*(s: PostgresDriver):
+                       Future[ArchiveDriverResult[seq[ArchiveRow]]] {.async.} =
+  ## Retrieve all messages from the store.
+  var rows: seq[Row]
+  var results: seq[ArchiveRow]
+  try:
+    rows = s.connection.getAllRows(sql("""SELECT storedAt, contentTopic,
+                                  payload, pubsubTopic, version, timestamp,
+                                  id FROM messages ORDER BY storedAt ASC"""))
+  except DbError:
+    return err("failed to query rows")
+
+  for r in rows:
+    let rowRes = extractRow(r)
+    if rowRes.isErr():
+      return err("failed to extract row")
+
+    results.add(rowRes.get())
+
+  return ok(results)
+
+method getMessages*(s: PostgresDriver,
+                    contentTopic: seq[ContentTopic] = @[],
+                    pubsubTopic = none(PubsubTopic),
+                    cursor = none(ArchiveCursor),
+                    startTime = none(Timestamp),
+                    endTime = none(Timestamp),
+                    maxPageSize = DefaultPageSize,
+                    ascendingOrder = true):
+                    Future[ArchiveDriverResult[seq[ArchiveRow]]] {.async.} =
+  var query = """SELECT storedAt, contentTopic, payload,
+  pubsubTopic, version, timestamp, id FROM messages"""
+  var statements: seq[string]
+  var args: seq[string]
+
+  if contentTopic.len > 0:
+    let cstmt = "contentTopic IN (" & "?".repeat(contentTopic.len).join(",") & ")"
+    statements.add(cstmt)
+    for t in contentTopic:
+      args.add(t)
+
+  if pubsubTopic.isSome():
+    statements.add("pubsubTopic = ?")
+    args.add(pubsubTopic.get())
+
+  if cursor.isSome():
```

Comment:
Can you please explain what this does?

Reply:
Sure! The concept of "cursor" is part of the pagination feature when retrieving messages. 

The client can optionally provide with a "PagingInfo" in the store protocol request that will therefore split the whole list of messages in pages (if page size is smaller than the # of messages retrieved after applying the search conditions.) 

Therefore, if a store response represents a page, that response will include a cursor that can be used to retrieve the next page. For the last page, the cursor is empty in the response.

An example of this can be seen at:
https://github.com/waku-org/nwaku/blob/ffac77611cd244980fbb98e24627186fb5f54d63/tests/v2/wakunode_rest/test_rest_store.nim#L219

Reply:
thanks!